crates/wasmtime/src/runtime/vm/cow.rs

```diff
@@ -470,19 +470,27 @@ impl MemoryImageSlot {
     /// process's memory on Linux. Up to that much memory will be `memset` to
     /// zero where the rest of it will be reset or released with `madvise`.
     #[allow(dead_code)] // ignore warnings as this is only used in some cfgs
-    pub(crate) fn clear_and_remain_ready(&mut self, keep_resident: usize) -> Result<()> {
+    pub(crate) fn clear_and_remain_ready(
+        &mut self,
+        keep_resident: usize,
+        decommit: impl FnMut(*mut u8, usize),
```

Comment:
Hm actually on second thought I have a different idea. The duplication in the three deallocate methods in `pooling.rs` is pretty subtle and feels sort of hard to get right. It also has the subtelty of the `Mutex<DeallocationQueue>` is held across function calls, such as this one, and it's getting to the point where it's easy to get into issues of lock ordering issues or deadlocks. For example in `deallocate_fiber_stack` the deallocation queue lock may be held across `self.stacks.deallocate` which internally acquires a different lock.

This leads me to a different way to model this which might be a bit less error-prone. This `decommit` argument could be replaced by a structure with `SmallVec`s inside of it where it's pushed onto. Once this step has completed there's a general method for "clear out this thing with small vectors or, if empty, do the deallocation immediately" or something like that. That would keep the critical section as small as possible while I think also making it a bit clearer how to reuse/shape code using it perhaps?

Reply:
I think the least-code way to do what you're suggesting is to pass an empty `DecommitQueue` into e.g. `image.clear_and_remain_ready` and unconditionally call `push_memory` on it; then pass it to a common method.

That method should check if its raw iovec queue is empty, and if so, immediately flush it, which will call `self.memories.deallocate` on the only memory pushed into it. Otherwise, take the lock and append its contents to the global `DecommitQueue`, and then if that queue is full, immediately flush that.

Having the ability to append one `DecommitQueue` into another might be nice for maintaining thread-local queues or something too, someday.

Also it might make sense to move the `batch_size` field to `PoolingInstanceAllocator`. We don't need to check it during `enqueue_raw`, especially with this proposed change, so we might as well store it alongside all the other data we need access to during flush.

Reply:
Sounds good to me 👍